### PR TITLE
Add initial AVX512 support for dot product on Linux

### PR DESCRIPTION
**Pull request by Ameobea**
NOTE: I am seeing different outputs when running with these changes.  They seem of equal quality, but this isn't something I observed when first testing this out on alpaca.cpp.

It's possible that some rounding behavior is happening slightly differently or something like that.  If this is a dealbreaker, I can try to figure out what is causing the difference and check if it's possible to get rid of it.

## Changes

 * Update Makefile to detect AVX512 support and add compiler flags if it's available
 * Add AVX512 impl based on existing AVX2 implementation, dot product on one 32-value block of 4-bit quantized ints at a time
 * Perform 8 bit -> 16 bit sign extension and multiply+add on 32 values at time instead of 16
 * Use built-in AVX512 horizontal reduce add to get sum at the end
 * Manual unrolling on inner dot product loop to reduce loop counter overhead

## Performance Impact

I initially implemented this [over on alpaca.cpp](https://github.com/antimatter15/alpaca.cpp/pull/80) where I saw an ~10% speedup to inferrence.

Before:

```
main: mem per token = 14368644 bytes
main:     load time =   923.25 ms
main:   sample time =    85.94 ms
main:  predict time = 23502.37 ms / 92.17 ms per token
main:    total time = 24845.69 ms
```

After:

```
main: mem per token = 14368644 bytes
main:     load time =   928.89 ms
main:   sample time =    16.18 ms
main:  predict time =  5720.41 ms / 82.90 ms per token
main:    total time =  6982.89 ms
```

I was hoping for more, but some other stuff I tried like converting the `bytesFromNibbles` function to operate on two blocks at a time by using AVX512 were not successful.

### CODE REVIEW

File: ggml.c
Comment (sw) [inline ggml.c:1421]: This should only be defined once (`#if defined(__AVX2__) || defined(__AVX512F__)`)

Comment (sw) [reply ggml.c:1421]: Should we be using nested functions?

Comment (Ameobea) [reply ggml.c]: Good call on re-using the `byteFromNibbles`.  Code has been updated.

As for nested functions, I'm not sure what you mean.  Do you mean the use of `process_one_block`?  I originally considered implementing it as a macro, but that made the code very messy with all the backslashes.

I also tried using `__attribute__((always_inline))` to make extra sure it always inlines, but that produced warnings.  I'm not super familiar with C/C++, so if there is a better way to achieve that I'm happy to use it.

That being said, this is always inlining for me and it would be very surprising if it were to fail to inline for pretty much any compiler, as long as optimizations are enabled.

Comment (sw) [reply ggml.c]: `process_one_block` is nested inside `ggml_vec_dot_q4_0`, that's not really standard C.

`ggml_vec_dot_q4_0` is at a point now where it should probably be refactored into a specific function for each kind of optimization.

Comment (Ameobea) [reply ggml.c]: Ah, I see.  I wasn't aware that wasn't standard in C.

I've split it out and renamed it to make it apparent that it's used as a helper function for the `dot_q4_0` function.

File: ggml.c
Comment (sw) [inline ggml.c:1275]: Pointers galore ;-)

I think this can made quite a bit simpler, but maybe that's for a separate PR.

The block size `bs` is fixed by `QK`, and you shouldn't really need separate pointers for the scale and the q4 values.

I'm thinking something like this:
```c
typedef struct {
  float scale;
  uin8_t q4[QK/2];
} q4_block;
```
Then just pass around two pointers of `*q4_block` for x and y.

Comment (Ameobea) [reply ggml.c:1275]: I tried to leave it as close as possible to the way it was before, just splitting it out to facilitate the manual unrolling with two accumulators to help with instruction pipelining and floating point numerical precision.

The start of the `ggml_vec_dot_q4_0` function defines those two different points for scales and quantized values separately, so I just keep using those instead of dong my own pointer arethmetic.

That being said, I'm not attached to the particulars of the arguments or layout or whatever.  If you or some other maintainer has some preference for how it's handled, feel free to just push to this branch with changes or let me know what you want me to change.

Comment (sw) [reply ggml.c:1275]: That's ok for now. I don't have AVX512 hardware though, so maybe someone else can have a look at this before merging it.

**Review by sw — APPROVED** [review]

**Comment by congdm** [comment]
Thank you for your effort! It also works on Windows and gives a little boost on my i7-11700F, from ~208 ms/token to 195 ms/token or sometimes even 185 ms/token on Alpaca7B.

**Pull request merged**
